File: aoc/2021/3.py
from bisect import bisect_left


def read_input(file):
    with open(file, "r") as f:
        lines = f.readlines()
    lines = [l.strip() for l in lines]
    return lines


def part_1(numbers):
    col_sum = [0] * len(numbers[0])

    for number in numbers:
        for i, c in enumerate(number):
            if c == "1":
                col_sum[i] += 1

    gamma_rate = ["1" if col > len(numbers) // 2 else "0" for col in col_sum]
    epsilon_rate = [str(1 - int(x)) for x in gamma_rate]

    power = int("".join(gamma_rate), base=2) * int("".join(epsilon_rate), base=2)

    return power


def part_2(numbers):
    """
    Sort the input
    Use binary search to find the first 1 in column
        - Determine whether 1 is the most common or least common bit
        - Update the left and right of the array and move on
        - The last remaining index is the index of the number to find
    """

    oxy_range = (0, len(numbers) - 1)
    co2_range = (0, len(numbers) - 1)

    bits = [[int(number[i]) for number in numbers] for i in range(len(numbers[0]))]

    def selection(numbers, valid_idx, col, oxy=True):

        if len(valid_idx) == 1:
            return valid_idx

        index_0 = []
        index_1 = []

        for idx in valid_idx:
            if numbers[idx][col] == "1":
                index_1.append(idx)
            else:
                index_0.append(idx)
        count_0 = len(index_0)
        half = len(valid_idx) // 2

        if oxy:
            # 1 is the most common bit
            if count_0 <= half:
                return index_1
            # 0 is the most common bit
            else:
                return index_0
        else:
            # 0 is the least common bit
            if count_0 <= half:
                return index_0
            # 1 is the least common bit
            else:
                return index_1

    oxy_idx = [i for i in range(len(numbers))]
    co2_idx = [i for i in range(len(numbers))]

    for i in range(len(numbers[0])):
        oxy_idx = selection(numbers, oxy_idx, i, True)
        co2_idx = selection(numbers, co2_idx, i, False)

    oxy_rating = int("".join(numbers[oxy_idx[0]]), base=2)
    co2_rating = int("".join(numbers[co2_idx[0]]), base=2)

    return oxy_rating * co2_rating


if __name__ == "__main__":
    numbers = read_input("./input.txt")
    print(part_2(numbers))

File: aoc/2021/3.cc
#include <bits/stdc++.h>
#include <fstream>
#include <string>

using namespace std;

long long SolvePartOne(const vector<string> &numbers) {
    int bit_length = numbers[0].length();
    vector<int> col_sum(bit_length, 0);

    for (auto &number : numbers) {
        for (int i = 0; i < (int)number.length(); i++) {
            auto c = number[i];
            if (c == '1') {
                col_sum[i] += 1;
            }
        }
    }
    vector<char> gamma_rate(bit_length, '0');
    vector<char> epsilon_rate(bit_length, '0');

    int threshold = (int)numbers.size() / 2;

    for (int i = 0; i < bit_length; i++) {
        if (col_sum[i] > threshold)
            gamma_rate[i] = '1';
        if (gamma_rate[i] == '0')
            epsilon_rate[i] = '1';
    }

    auto ToDecimal = [](const vector<char> &bits) {
        string bit_str(bits.begin(), bits.end());
        return stoi(bit_str, 0, 2);
    };
    return ToDecimal(gamma_rate) * ToDecimal(epsilon_rate);
}

long long SolvePartTwo(const vector<string> &numbers) {
    pair<int, int> oxy_range(0, (int)numbers.size() - 1);
    pair<int, int> co2_range(0, (int)numbers.size() - 1);

    int bit_length = numbers[0].length();
    int n = numbers.size();
    vector<vector<int>> bits(bit_length, vector<int>(n));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < bit_length; j++) {
            bits[j][i] = (int)(numbers[i][j] - '0');
        }
    }

    auto selection = [](const vector<string> &numbers,
            const vector<int> &valid_idx, int col, bool oxy) {
        if (valid_idx.size() == 1)
            return valid_idx;

        vector<int> index_0, index_1;

        for (auto idx : valid_idx) {
            if (numbers[idx][col] == '1')
                index_1.push_back(idx);
            else
                index_0.push_back(idx);
        }
        int zero_count = index_0.size();
        int half = valid_idx.size() / 2;

        if (oxy) {
            if (zero_count <= half)
                return index_1;
            else
                return index_0;
        } else {
            if (zero_count <= half)
                return index_0;
            else
                return index_1;
        }
    };

    vector<int> oxy_idx, co2_idx;

    for (int i = 0; i < n; i++) {
        oxy_idx.push_back(i);
        co2_idx.push_back(i);
    }

    for (int i = 0; i < bit_length; i++) {
        oxy_idx = selection(numbers, oxy_idx, i, true);
        co2_idx = selection(numbers, co2_idx, i, false);
    }

    auto ToDecimal = [](const vector<char> &bits) {
        string bit_str(bits.begin(), bits.end());
        return stoi(bit_str, 0, 2);
    };

    vector<char> oxy_rating, co2_rating;

    for (int i = 0; i < bit_length; i++) {
        oxy_rating.push_back(numbers[oxy_idx[0]][i]);
        co2_rating.push_back(numbers[co2_idx[0]][i]);
    }

    return ToDecimal(oxy_rating) * ToDecimal(co2_rating);
}

int main(int argc, char *argv[]) {
    ifstream fin("input.txt", fstream::in);
    vector<string> numbers;
    string line;
    while (fin >> line) {
        numbers.push_back(line);
    }
    cout << SolvePartOne(numbers) << endl;
    cout << SolvePartTwo(numbers) << endl;

    return 0;
}
